pest/src/inputs/string_input.rs

```diff
@@ -30,10 +36,182 @@ impl StringInput {
     /// assert_eq!(input.len(), 3);
     /// ```
     pub fn new(string: String) -> StringInput {
-        StringInput {
-            string
+        StringInput { string }
+    }
+}
+
+impl<'a> StrInput<'a> {
+    /// Creates a new `StrInput` from a `&str`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest::inputs::{Input, StrInput};
+    /// let input = StrInput::new("asd");
+    ///
+    /// assert_eq!(input.len(), 3);
+    /// ```
+    pub fn new(source: &'a str) -> StrInput<'a> {
+        StrInput { str_ref: source }
+    }
+}
+
+#[inline]
+unsafe fn line_col(source: &str, pos: usize) -> (usize, usize) {
+    if pos > source.len() {
+        panic!("position out of bounds");
+    }
+
+    let mut pos = pos;
+    let slice = &source[..pos];
+    let mut chars = slice.chars().peekable();
+
+    let mut line_col = (1, 1);
+
+    while pos != 0 {
+        match chars.next() {
+            Some('\r') => {
+                if let Some(&'\n') = chars.peek() {
+                    chars.next();
+
+                    if pos == 1 {
+                        pos -= 1;
+                    } else {
+                        pos -= 2;
+                    }
+
+                    line_col = (line_col.0 + 1, 1);
+                } else {
+                    pos -= 1;
+                    line_col = (line_col.0, line_col.1 + 1);
+                }
+            }
+            Some('\n') => {
+                pos -= 1;
+                line_col = (line_col.0 + 1, 1);
+            }
+            Some(c) => {
+                pos -= c.len_utf8();
+                line_col = (line_col.0, line_col.1 + 1);
+            }
+            None => unreachable!(),
         }
     }
+
+    line_col
+}
+
+#[inline]
+unsafe fn line_of(source: &str, mut pos: usize) -> &str {
+    if pos > source.len() {
+        panic!("position out of bounds");
+    }
+
+    let start = if pos == 0 {
+        0
+    } else {
+        if source.slice_unchecked(pos, pos + 1) == "\n" {
+            pos -= 1;
+        }
+
+        let start = source.char_indices()
+                          .rev()
+                          .skip_while(|&(i, _)| i > pos)
+                          .find(|&(_, c)| c == '\n');
+        match start {
+            Some((i, _)) => i + 1,
+            None => 0,
+        }
+    };
+
+    let end = if source.len() == 0 {
+        0
+    } else if pos == source.len() - 1 {
+        let mut end = source.len();
+
+        if end > 0 && source.slice_unchecked(end - 1, end) == "\n" {
+            end -= 1;
+        }
+        if end > 0 && source.slice_unchecked(end - 1, end) == "\r" {
+            end -= 1;
+        }
+
+        end
+    } else {
+        let end = source.char_indices().skip_while(|&(i, _)| i < pos).find(
```

Comment:
This should be formatted like the one above.

Reply:
@dragostis I'll suggest you to put rustfmt.toml under the root directory of this project. So contributors get same rustfmt configuration and we can keep the code consistent. 

And for these formation issues, we can leave it for rustfmt iterations in future

Reply:
I really considered that, but formatting can become really poor in some particular cases, like the one above, and it's not even consistent with the prior example. Putting the file there pretty much means keeping it. Right now, using rustfmt on the whole project makes the formatter hang because of an issues with nested lambdas. I am very much in favor of going down this road, but right now I would say it's partly too much of a hassle, and partly making the code messy like the example above.

Reply:
I see. I had issue with saving `pest/test/json.rs` and thought it was issue with Emacs' rust-mode.